src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+}
+
+static inline void *
+platform_heap_id_to_shmaddr(platform_heap_id hid)
+{
+   debug_assert(hid != NULL);
+   void *shmaddr = (void *)platform_heap_id_to_handle(hid);
+   return shmaddr;
+}
+
+/* Evaluates to valid 'low' address within shared segment. */
+static inline void *
+platform_shm_lop(platform_heap_id hid)
+{
+   return (void *)platform_heap_id_to_handle(hid);
+}
+
+/*
+ * Evaluates to valid 'high' address within shared segment.
+ * shm_end points to one-byte-just-past-end of shared segment.
+ * Valid 'high' byte is last byte just before shm_end inside segment.
+ */
+static inline void *
+platform_shm_hip(platform_heap_id hid)
+{
+   return (((shmem_info *)platform_heap_id_to_shmaddr(hid))->shm_end - 1);
+}
+
+static inline void
+shm_lock_mem_frags(shmem_info *shminfo)
+{
+   platform_spin_lock(&shminfo->shm_mem_frags_lock);
+}
+
+static inline void
+shm_unlock_mem_frags(shmem_info *shminfo)
+{
+   platform_spin_unlock(&shminfo->shm_mem_frags_lock);
+}
+
+/*
+ * Address 'addr' is valid if it's just past end of control block and within
+ * shared segment.
+ */
+static inline bool
+platform_valid_addr_in_shm(platform_heap_handle heap_handle, void *addr)
+{
+   debug_assert(platform_shm_heap_handle_valid(heap_handle),
+                "Shared memory heap_handle %p is invalid.\n",
+                heap_handle);
+
+   const shmem_info *shmaddr = (shmem_info *)heap_handle;
+   return ((addr >= ((void *)shmaddr + platform_shm_ctrlblock_size()))
+           && (addr < shmaddr->shm_end));
+}
+
+/*
+ * platform_valid_addr_in_heap(), platform_valid_addr_in_shm()
+ *
+ * Validate that input address 'addr' is a valid address within shared segment
+ * region.
+ */
+static inline bool
+platform_valid_addr_in_heap(platform_heap_id heap_id, void *addr)
+{
+   return platform_valid_addr_in_shm(platform_heap_id_to_handle(heap_id), addr);
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shmcreate() -- Create a new / attach to an existing shared segment.
+ *
+ * For a given set of heap ID/handle, we expect that this create method will
+ * only be called once. [ Otherwise, it means some code is erroneously creating
+ * the shared segment twice, clobbering previously established handles. ]
+ * -----------------------------------------------------------------------------
+ */
+platform_status
+platform_shmcreate(size_t                size,
+                   platform_heap_handle *heap_handle, // Out
+                   platform_heap_id     *heap_id)         // Out
+{
+   platform_assert((*heap_handle == NULL),
+                   "Heap handle is expected to be NULL while creating a new "
+                   "shared segment.\n");
+   platform_assert((*heap_id == NULL),
+                   "Heap handle is expected to be NULL while creating a new "
+                   "shared segment.\n");
+
+   int shmid = shmget(0, size, (IPC_CREAT | PLATFORM_IPC_OBJS_PERMS));
```

Comment:
@rosenhouse wanted me to implement this with System-V IPC APIs. I didn't get around to that.

I'd like this to stabilize and integrate. Will re-visit this aspect in part-3 PR while reviews for part-1 and part-2 are on-going. I don't think this is a big blocker issue right now. 

Let me know your opinions.

Reply:
@rtjohnso & I re-reviewed this in closing meeting. For now, as this functionality works, and it allows us someday to be able to do `shmattach()` -- we can live with this.

Nothing further to do at this time.